src/main/scala/mesosphere/marathon/core/deployment/impl/TaskReplaceActor.scala

```diff
@@ -80,13 +80,27 @@ class TaskReplaceActor(
     instancesByIncarnation._1.sortBy(_.appTask.status.stagedAt) ++ instancesByIncarnation._2
   }
 
+  // Return the number of non running old instances killed
+  def killNonRunningOldInstances(oldInstances: Seq[Instance]): Integer = {
+    val nonRunningOldInstances = oldInstances.filter(!_.isRunning).to[mutable.Queue]
+    val nonRunningOldInstancesCount = nonRunningOldInstances.size
+    for (_ <- 0 until nonRunningOldInstancesCount) killNextOldInstance(nonRunningOldInstances)
+    return nonRunningOldInstancesCount
+  }
+
   def step(health: Map[Instance.Id, Seq[Health]]): Unit = {
     logger.debug(s"---=== DEPLOYMENT STEP FOR ${pathId} ===---")
     val current_instances = instanceTracker.specInstancesSync(pathId).partition(_.runSpecVersion == runSpec.version)
 
     val new_instances = current_instances._1.partition(isHealthy(_, health))
     val old_instances = current_instances._2.partition(isHealthy(_, health))
 
+    if (killNonRunningOldInstances(old_instances._2) > 0) {
+      logger.info("Found and killed non running instances from a previous, likely failing, deployment. Was a new deployment applied forcefully?")
+      logger.info("Aborting current deployment step and waiting for the next one.")
+      return
+    }
+
     // make sure we kill in order:
     // - instances with unhealthy tasks
     // - instances with oldest tasks
```

Comment:
Can't put a comment but on L94, maybe we should filter toKill to contain only running instances to avoid send two times the kill signal to the same non running app?

Reply:
Nah sorry, just remember you abort the step() so it won't happen ! Resolved

Reply:
When non running old instances are killed at step n, we abort step n and wait for step n+1 before trying to kill other old instances. So I assume instances killed at step n will be removed from the instance tracker, and that toKill will only contain running old instances.
Or am I missing something?

Reply:
Yup no I saw that after, this is why I put a new comment and resolved. PR is fine, you can merge :)